Refracting to TypeScript and Vite

tanadendere opened the pull request:
This pull request introduces transitioning the codebase from JavaScript to TypeScript for improved type safety and maintainability. Also, the build process has been revamped to utilise Vite as the building tool, to enhance performance and developer experience

tanadendere committed:
feat: :sparkles: Refracted project to use Vite

Created a new Vite project then consolidated the created files with the project's old files

tanadendere committed:
feat: :construction: Changed .js file extensions to .ts

Refractored all the file extensions to be typecript

tanadendere committed:
feat: :sparkles: Typed the Leaflet map

Installed the leaflet npm package and used it to type the map functionalities

tanadendere committed:
feat: :art: Separation of map concerns

Moved map listeners to map script file

tanadendere committed:
feat: :construction: Typed the index.ts file

Typed all the functions in index.ts as well as declaring variables as const

tanadendere committed:
feat: :sparkles: Made new forecast type

Created a forecast interface and used it to type the function that calls the api

tanadendere committed:
refactor: :construction: Finished typing the app

The whole weather app is now typed

yanevdw commented on tsconfig.json:
```diff
@@ -0,0 +1,23 @@
+    "strict": true,
+    "noUnusedLocals": true,
+    "noUnusedParameters": true,
+    "noFallthroughCasesInSwitch": true
```
I would suggest adding the following rules:

```
  "noImplicitAny" : true,
  "strictFunctionTypes" : true,
  "noImplicitThis" : true,
  "alwaysStrict" : true,
  "strictPropertyInitialization": true
```

Meggie-H commented on src/scripts/index.ts:
```diff
@@ -0,0 +1,75 @@
+
+for (let city of majorCities) {
+  city.addEventListener("click", function (e) {
+    console.log(city.innerText);
```
Remember to take you console.logs out

theadley replied:
Please remove

Meggie-H commented on src/scripts/index.ts:
```diff
@@ -0,0 +1,75 @@
+export function callTheWeatherAPI(
+  longitude: number,
+  latitude: number,
+  city: string
```
I see you use longitude, latitude and city together quite often. You could maybe create an interface for it and call it cityData or something in that line.

Meggie-H commented on src/scripts/dom.ts:
```diff
@@ -27,43 +29,47 @@ function setTodayElement(todays_data, city) {
   // weather icon
-  let icon = document.getElementById("weather_icon");
-  icon.src = getIconSrc(todays_data.weather);
+  const icon: HTMLElement | null = document.getElementById("weather_icon");
```
You do not have to specify the type as it is implied by `getElementById`.

AshirMohab commented on src/scripts/index.ts:
```diff
@@ -0,0 +1,75 @@
+      // mapping of results
+      console.log(res);
+      console.log("---");
+      setDOM(res, city);
```
remove the `console.logs`

AshirMohab commented on src/scripts/map.ts:
```diff
@@ -0,0 +1,59 @@
+    let longitude = Math.round(e.latlng.lat * 1000) / 1000;
+    let latitude = Math.round(e.latlng.lng * 1000) / 1000;
+    // console.log(longitude);
+    // console.log(latitude);
```
same here

yanevdw commented on src/scripts/dom.ts:
```diff
@@ -77,156 +83,132 @@ function getDayOfWeek(num) {
-    let forecast_element = setDay(day);
-    seven_day_forecast.appendChild(forecast_element);
-  });
+function set7dayForecast(forecast_data: Series[]): void {
+  let seven_day_forecast: HTMLElement | null =
+    document.getElementById("seven_day_forecast");
+  if (seven_day_forecast) {
```
Instead of setting the `seven-day-forecast-element` to be either of type `HTMLElement | null` I would adives you to do this:
```ts
const seven_day_forecast = document.getElementById("seven-day-forecast");
// This is a guard clause
if (!seven_day_forecast) {
  return;
}

// Then do the rest of your code for when seven_day_forecast exists (is not null).
  
```

yanevdw commented on src/scripts/index.ts:
```diff
@@ -0,0 +1,75 @@
+});
+
+// map listeners
+const map_div: HTMLElement | null = document.getElementById("map_container");
+const seven_day: HTMLElement | null =
+  document.getElementById("seven_day_forecast");
+const open_map: HTMLElement | null =
+  document.getElementById("map_button_float");
+
+if (map_div && seven_day && open_map) {
+  map_listeners(map, map_div, seven_day, open_map);
+}
```
As above, I would rather not let the types possibly be null.
```ts
const map_div = document.getElementById("map_container");
const seven_day = document.getElementById("seven_day_forecast");
const open_map = document.getElementById("map_button_float");

if (!map_div || seven_day || open_map) {
  //Handle what happens when these are null. For example, throw an error (this would require a try/catch.
}

map_listeners(map, map_div, seven_day, open_map);
```

yanevdw commented on src/scripts/map.ts:
```diff
@@ -0,0 +1,59 @@
+  latitude: number,
+  city: string
+): void {
+  console.log("changing map", city);
```
Please don't use console.log.

yanevdw replied:
Please don't console.log().

yanevdw reviewed (changes requested):
Minor changes suggested. Please take note of TypeScript naming conventions as well, which uses camelCase for both function and variable names.

tanadendere committed:
refactor: :construction: Removed console.logs

Improvements from PR comments

theadley commented on package.json:
```diff
@@ -0,0 +1,20 @@
+  },
+  "dependencies": {
+    "@types/geojson": "^7946.0.14",
+    "@types/leaflet": "^1.9.8",
```
Types are dev dependencies, please move them up

theadley commented on src/scripts/dom.ts:
```diff
@@ -10,10 +12,10 @@ export function setDOM(results, city) {
   set7dayForecast(results.dataseries);
```
Don't leave commented out code nor console.log in your codebase please

theadley commented on src/scripts/dom.ts:
```diff
@@ -10,10 +12,10 @@ export function setDOM(results, city) {
   // todays date
-  let todays_date = getDate(todays_data.date.toString());
-  const monthNames = [
+  const todays_date: Date = getDate(todays_data.date.toString());
```
🐪 >> 🐍

theadley commented on src/scripts/dom.ts:
```diff
@@ -27,43 +29,47 @@ function setTodayElement(todays_data, city) {
-  todayElement.innerText = datestring;
 
+  const todayElement: HTMLElement | null =
+    document.getElementById("todaysdate");
```
Name the element with kebab-case
i.e. `todays-date`

theadley commented on src/scripts/dom.ts:
```diff
@@ -27,43 +29,47 @@ function setTodayElement(todays_data, city) {
-  let location = document.getElementById("location_name");
-  location.innerText = city;
-
+  const location: HTMLElement | null = document.getElementById("location_name");
```
🐍

theadley commented on src/scripts/dom.ts:
```diff
@@ -27,43 +29,47 @@ function setTodayElement(todays_data, city) {
-  );
-
+  const temp_ranges: HTMLElement | null =
+    document.getElementById("temp_ranges");
```
🐍

theadley replied:
🍡

theadley commented on src/scripts/dom.ts:
```diff
@@ -27,43 +29,47 @@ function setTodayElement(todays_data, city) {
-  icon.src = getIconSrc(todays_data.weather);
+  const icon: HTMLElement | null = document.getElementById("weather_icon");
+
+  if (todayElement && location && temp_ranges && weather && icon) {
```
💂🏻

theadley commented on src/scripts/dom.ts:
```diff
@@ -77,156 +83,132 @@ function getDayOfWeek(num) {
-    seven_day_forecast.appendChild(forecast_element);
-  });
+function set7dayForecast(forecast_data: Series[]): void {
+  let seven_day_forecast: HTMLElement | null =
```
No reassign? No let

theadley commented on src/scripts/dom.ts:
```diff
@@ -77,156 +83,132 @@ function getDayOfWeek(num) {
+  if (seven_day_forecast) {
+    seven_day_forecast.innerHTML = "";
+
+    for (let day of forecast_data) {
```
const as you do not reassign day within the scope of the loop

theadley commented on src/scripts/dom.ts:
```diff
@@ -77,156 +83,132 @@ function getDayOfWeek(num) {
-function setDay(days_data) {
-  const forecast_item = document.createElement("div");
+function setDay(days_data: Series): HTMLDivElement {
+  let forecast_item: HTMLDivElement = document.createElement("div");
```
let or const?

theadley commented on src/scripts/dom.ts:
```diff
@@ -77,156 +83,132 @@ function getDayOfWeek(num) {
-  const forecast_item = document.createElement("div");
+function setDay(days_data: Series): HTMLDivElement {
+  let forecast_item: HTMLDivElement = document.createElement("div");
   forecast_item.className = "forecast_item";
```
🍡

theadley commented on src/scripts/dom.ts:
```diff
@@ -77,156 +83,132 @@ function getDayOfWeek(num) {
   // day
-  let day_date = getDate(days_data.date.toString());
-  let day = getDayOfWeek(day_date.getDay());
+  const day_date: Date = getDate(days_data.date.toString());
```
🐪
Type is inferred by the return type of the function

theadley commented on src/scripts/index.ts:
```diff
@@ -0,0 +1,75 @@
+
+const majorCities: HTMLElement[] = Array.from(getMajorCities);
+
+for (let city of majorCities) {
```
const

theadley commented on src/scripts/index.ts:
```diff
@@ -0,0 +1,75 @@
+  });
+}
+
+async function getCoords(city: string): Promise<void> {
```
Why not in api.ts?

theadley commented on src/scripts/index.ts:
```diff
@@ -0,0 +1,75 @@
+    })
+    .then((data) => {
+      // returns the coordinates of the city
+      const longitude: number = data.major_cities[0][city][0].long;
```
Safety checks please. Don't trust the API, it's what _they_ want you to do.

theadley commented on src/scripts/index.ts:
```diff
@@ -0,0 +1,75 @@
+      updateMap(map, longitude, latitude, city);
+    })
+    .catch(console.error)
+    .finally(() => {
```
You can omit unless you need it to hide spinners or do cleanup, etc.

theadley commented on src/scripts/map.ts:
```diff
@@ -0,0 +1,59 @@
+): void {
+  console.log("changing map", city);
+
+  map.panTo(latLng(latitude, longitude));
```
`flyTo` is a cool animated version of this if you want

theadley commented on src/scripts/map.ts:
```diff
@@ -0,0 +1,59 @@
+  seven_day: HTMLElement,
+  open_map: HTMLElement
+): void {
+  open_map.addEventListener("click", function (e) {
```
Don't do DOM manipulation in the map. Take a callback function in.
Use the orchestrator to hand in a DOM function in the callback param here.

theadley reviewed (changes requested):
🐍 bad, 🍡 & 🐪 good💂🏻, const, trivially inferred types.
No commented out code nor console.logs please

tanadendere committed:
refactor: :art: Naming conventions and adding Linting configurations

Changed all the names of functions, variables, ids, etc.. to fit the naming convections. And added Eslint as well as recommended Linting configurations for TS

tanadendere committed:
feat: :sparkles: Added a city data interface

Created a city data interface that has the values of longitude, latitude and city as they are always used together

tanadendere committed:
refactor: :art: Added guard clauses and did further separation of concerns

tanadendere committed:
fix: :sparkles: Added guard clause for weather api

AshirMohab commented on src/scripts/dom.ts:
```diff
@@ -0,0 +1,244 @@
+  // todays date
+  const todaysDate: Date = getDate(todaysWeatherData.date.toString());
+  const monthNames: string[] = [
+    "January",
```
I mentioned to Jules but you can move the `monthNames` and `days` to their own file. They are static and can just be imported where used.
